File: apps/windows/Wisp.ProjectBrowser.Contracts/NativeResearchArchive.cs
using System.Text.Json;
using System.Text.Json.Nodes;
namespace Wisp.ProjectBrowser.Contracts;

public sealed record NativeArchiveScript(string Filename, string Content);
public sealed record NativeArchiveFile(string Path, string Checksum, ulong SizeBytes, string Action,
    bool CanDelete, string Reason, string? SnapshotPath, string CleanupStatus);
public sealed record NativeArchiveFileChoice(string Path, string Action);
public sealed record NativeArchiveConfirmation(string Id, string Title, string Report, NativeArchiveScript[] Scripts, NativeArchiveFileChoice[] Files);
public sealed record NativeResearchArchive(string Id, string ProjectId, string FrameId, string SourceHash,
    string Title, string Report, NativeArchiveScript[] Scripts, NativeArchiveFile[] Files,
    long CreatedAt, long? FrozenAt, string[] Warnings)
{
    public static NativeResearchArchive? Decode(JsonNode? node, string projectId, string sessionId)
    {
        if (node is null) return null;
        var result = node.Deserialize<NativeResearchArchive>(ConversationSnapshot.JsonOptions)
            ?? throw new InvalidDataException("Missing research archive");
        if (result.ProjectId != projectId || result.FrameId != sessionId || string.IsNullOrEmpty(result.Id)
            || result.Scripts is null || result.Files is null || result.Warnings is null)
            throw new InvalidDataException("Archive identity mismatch");
        return result;
    }
    public NativeArchiveConfirmation Confirmation() => new(Id, Title, Report, Scripts, Files.Select(f => new NativeArchiveFileChoice(f.Path, f.Action)).ToArray());
}
/// <summary>Do not replay archive mutations after an ambiguous failure. Re-read
/// the saved record; a failed cleanup can follow a successful immutable freeze.</summary>
public interface INativeArchiveClient
{
    Task<NativeResearchArchive?> GetAsync(string projectId, string sessionId, CancellationToken cancellationToken = default);
    Task<NativeResearchArchive?> PrepareAsync(string projectId, string sessionId, CancellationToken cancellationToken = default);
    Task<NativeResearchArchive?> ConfirmAsync(string projectId, string sessionId, NativeArchiveConfirmation input, CancellationToken cancellationToken = default);
    Task<NativeResearchArchive?> RetryCleanupAsync(string projectId, string sessionId, CancellationToken cancellationToken = default);
    Task<string> ContinueAsync(string projectId, string sessionId, CancellationToken cancellationToken = default);
}
public sealed class NativeArchiveClient(INativeSettingsClient transport) : INativeArchiveClient
{
    private async Task<NativeResearchArchive?> Call(string action, string project, string session, NativeArchiveConfirmation? input, CancellationToken token)
    {
        JsonObject args = new() { ["session_id"] = session };
        if (input is not null) args["input"] = JsonSerializer.SerializeToNode(input, ConversationSnapshot.JsonOptions);
        return NativeResearchArchive.Decode(await transport.InvokeAsync("native_conversation_archive_" + action, args, project, token).ConfigureAwait(false), project, session);
    }
    public Task<NativeResearchArchive?> GetAsync(string projectId, string sessionId, CancellationToken cancellationToken = default) => Call("get", projectId, sessionId, null, cancellationToken);
    public Task<NativeResearchArchive?> PrepareAsync(string projectId, string sessionId, CancellationToken cancellationToken = default) => Call("prepare", projectId, sessionId, null, cancellationToken);
    public Task<NativeResearchArchive?> ConfirmAsync(string projectId, string sessionId, NativeArchiveConfirmation input, CancellationToken cancellationToken = default) => Call("confirm", projectId, sessionId, input, cancellationToken);
    public Task<NativeResearchArchive?> RetryCleanupAsync(string projectId, string sessionId, CancellationToken cancellationToken = default) => Call("retry", projectId, sessionId, null, cancellationToken);
    public async Task<string> ContinueAsync(string projectId, string sessionId, CancellationToken cancellationToken = default) =>
        (await transport.InvokeAsync("native_conversation_archive_continue", new() { ["session_id"] = sessionId }, projectId, cancellationToken).ConfigureAwait(false))?.GetValue<string>()
            ?? throw new InvalidDataException("Missing continuation ID");
}

File: apps/macos/Sources/WispProjectBrowser/NativeResearchArchive.swift
import Foundation

public struct NativeArchiveScript: Codable, Equatable, Sendable {
    public var filename: String
    public var content: String
}
public struct NativeArchiveFile: Codable, Equatable, Sendable {
    public let path: String
    public let checksum: String
    public let size_bytes: UInt64
    public var action: String
    public let can_delete: Bool
    public let reason: String
    public let snapshot_path: String?
    public let cleanup_status: String
}
/// Existing wisp-dto ResearchArchive. Secrets and file deletion remain in the host.
public struct NativeResearchArchive: Codable, Equatable, Sendable {
    public let id: String
    public let project_id: String
    public let frame_id: String
    public let source_hash: String
    public var title: String
    public var report: String
    public var scripts: [NativeArchiveScript]
    public var files: [NativeArchiveFile]
    public let created_at: Int64
    public let frozen_at: Int64?
    public let warnings: [String]
    public static func decode(_ value: SettingsValue, project: String, session: String) throws -> Self {
        let result = try JSONDecoder().decode(Self.self, from: JSONEncoder().encode(value))
        guard result.project_id == project, result.frame_id == session, !result.id.isEmpty else { throw ProjectBrowserError.invalidResponse }
        return result
    }
    public func confirmation() throws -> SettingsValue {
        var value = try JSONDecoder().decode(SettingsValue.self, from: JSONEncoder().encode(self))
        value = .object(["id": value["id"], "title": value["title"], "report": value["report"], "scripts": value["scripts"], "files": .array(files.map { .object(["path": .string($0.path), "action": .string($0.action)]) })])
        return value
    }
}
